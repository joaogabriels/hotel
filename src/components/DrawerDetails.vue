<template>
  <q-card flat>
    <q-card-section class="row items-center q-pb-none">
      <p class="text-h6 q-ma-none">{{ hotel.name }}</p>
    </q-card-section>

    <q-card-section class="q-pt-none">
      <p class="text-subtitle1 q-ma-none">{{ hotel.address.fullAddress }}</p>
    </q-card-section>

    <q-card-section class="q-pt-none">
      <div class="column q-gutter-y-md">
        <div class="carousel-container">
          <HotelImagesCarousel :images="hotel.images" :height="$q.screen.lt.md ? '256px' : '320px'" :arrows="true" />
        </div>

        <div class="text-body1">{{ hotel.description }}</div>

        <div v-if="hotel.amenities?.length">
          <div class="text-subtitle2 q-mb-sm">Comodidades</div>
          <HotelAmenities :amenities="hotel.amenities" show-labels />
        </div>
      </div>
    </q-card-section>
  </q-card>
</template>

<script setup lang="ts">
import type { Hotel } from './models';
import HotelAmenities from './HotelAmenities.vue';
import HotelImagesCarousel from './HotelImagesCarousel.vue';

defineProps<{
  hotel: Hotel;
}>();
</script>

<style>
.carousel-container {
  border-radius: 12px;
}
</style>