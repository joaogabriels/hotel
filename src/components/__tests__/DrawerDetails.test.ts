import { mount } from '@vue/test-utils'
import { describe, expect, it } from 'vitest'
import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-vitest'

import DrawerDetails from '../DrawerDetails.vue'
import { mockHotel } from './mocks'

installQuasarPlugin()

describe('DrawerDetails', () => {
  it('matches snapshot', () => {
    const wrapper = mount(DrawerDetails, { props: { hotel: mockHotel } })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it('should render hotel amenities when they exist', () => {
    const wrapper = mount(DrawerDetails, { props: { hotel: mockHotel } })

    const amenitiesSection = wrapper.find('.amenities-section')
    const amenitiesTitle = wrapper.findAll('.drawer-section-title')

    expect(amenitiesSection.exists()).toBe(true)
    expect(amenitiesTitle[0]?.text()).toBe('Facilidades do hotel')
  })

  it('should be able to expand amenities', async () => {
    const wrapper = mount(DrawerDetails, { props: { hotel: mockHotel } })

    const expandButton = wrapper.findAll('.q-btn')[2]

    await expandButton?.trigger('click')

    expect(expandButton?.text()).toBe('Ver menos')
  })

  it('should be able to collapse amenities', async () => {
    const wrapper = mount(DrawerDetails, { props: { hotel: mockHotel } })

    const expandButton = wrapper.findAll('.q-btn')[2]

    await expandButton?.trigger('click')
    await expandButton?.trigger('click')

    expect(expandButton?.text()).toBe('Mostrar todas as facilidades')
  })

  it('should be able to render hotel description', () => {
    const wrapper = mount(DrawerDetails, { props: { hotel: mockHotel } })

    const descriptionSection = wrapper.findAll('.drawer-section-title')[1]

    expect(descriptionSection?.exists()).toBe(true)
  })

  it('should emit close event when clicking close button on mobile', async () => {
    const wrapper = mount(DrawerDetails, {
      props: { hotel: mockHotel },
      global: {
        mocks: {
          $q: {
            screen: {
              lt: {
                md: true,
              },
            },
            platform: {
              has: {
                touch: false,
              },
            },
            lang: {
              rtl: false,
            },
          },
        },
      },
    })

    const closeButton = wrapper.find('.q-btn')
    await closeButton.trigger('click')

    expect(wrapper.emitted()).toHaveProperty('close')
    expect(wrapper.emitted('close')).toHaveLength(1)
  })
})
