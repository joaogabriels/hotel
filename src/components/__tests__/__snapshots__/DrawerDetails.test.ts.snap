// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`HotelDrawerDetails > matches snapshot 1`] = `
"<div data-v-69d547a9=\\"\\" class=\\"q-scrollarea fit\\">
  <div class=\\"q-scrollarea__container scroll relative-position fit hide-scrollbar\\">
    <div class=\\"q-scrollarea__content absolute\\">
      <!---->
      <div data-v-69d547a9=\\"\\" class=\\"q-pa-md\\">
        <div data-v-69d547a9=\\"\\" class=\\"row items-start justify-between q-mb-md\\">
          <div data-v-69d547a9=\\"\\" class=\\"col\\">
            <div data-v-378063dc=\\"\\" data-v-69d547a9=\\"\\" class=\\"row items-start q-gutter-x-xs\\"><span data-v-378063dc=\\"\\">5</span>
              <div data-v-378063dc=\\"\\" class=\\"q-rating row inline items-center q-rating--non-editable text-orange-4 rating\\" style=\\"font-size: 18px;\\" role=\\"radiogroup\\" aria-readonly=\\"true\\">
                <div class=\\"q-rating__icon-container flex flex-center\\" role=\\"radio\\" aria-checked=\\"false\\" aria-label=\\"grade 1\\"><i class=\\"q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-orange-4\\" aria-hidden=\\"true\\" role=\\"presentation\\">grade</i></div>
                <div class=\\"q-rating__icon-container flex flex-center\\" role=\\"radio\\" aria-checked=\\"false\\" aria-label=\\"grade 2\\"><i class=\\"q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-orange-4\\" aria-hidden=\\"true\\" role=\\"presentation\\">grade</i></div>
                <div class=\\"q-rating__icon-container flex flex-center\\" role=\\"radio\\" aria-checked=\\"false\\" aria-label=\\"grade 3\\"><i class=\\"q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-orange-4\\" aria-hidden=\\"true\\" role=\\"presentation\\">grade</i></div>
                <div class=\\"q-rating__icon-container flex flex-center\\" role=\\"radio\\" aria-checked=\\"false\\" aria-label=\\"grade 4\\"><i class=\\"q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-orange-4\\" aria-hidden=\\"true\\" role=\\"presentation\\">grade</i></div>
                <div class=\\"q-rating__icon-container flex flex-center\\" role=\\"radio\\" aria-checked=\\"true\\" aria-label=\\"grade 5\\"><i class=\\"q-icon notranslate material-icons q-rating__icon q-rating__icon--active text-orange-4\\" aria-hidden=\\"true\\" role=\\"presentation\\">grade</i></div>
              </div>
            </div>
            <p data-v-69d547a9=\\"\\" class=\\"text-h5 q-ma-none text-grey-9 q-mt-sm\\">Hotel de Teste</p>
            <p data-v-69d547a9=\\"\\" class=\\"text-subtitle2 text-grey-6 q-ma-none text-weight-regular\\">Av. Antônio Abrahão Caram, 1001, São José, Belo Horizonte, MG, Brasil, 31275-000</p>
          </div>
          <!--v-if-->
        </div>
        <div data-v-69d547a9=\\"\\" class=\\"carousel-container q-mb-md\\">
          <div data-v-f3901ebe=\\"\\" data-v-69d547a9=\\"\\" class=\\"q-carousel q-panel-parent q-carousel--without-padding hotel-carousel\\" style=\\"height: 420px;\\">
            <div class=\\"q-carousel__slides-container\\">
              <transition-stub appear=\\"false\\" persisted=\\"false\\" css=\\"true\\">
                <div class=\\"q-panel scroll\\" style=\\"--q-transition-duration: 300ms;\\" role=\\"tabpanel\\">
                  <div data-v-f3901ebe=\\"\\" class=\\"q-carousel__slide\\"><img data-v-f3901ebe=\\"\\" src=\\"https://fakeimg.pl/600x400\\" class=\\"full-width full-height fit\\"></div>
                </div>
              </transition-stub>
            </div>
            <div data-v-f3901ebe=\\"\\" class=\\"q-carousel__control absolute absolute-left carousel-nav-btn\\" style=\\"margin: 210px 16px;\\"><button data-v-f3901ebe=\\"\\" class=\\"q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--round bg-white text-black q-btn--actionable q-focusable q-hoverable q-btn--dense carousel-nav-btn\\" style=\\"font-size: 12px;\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><i class=\\"q-icon notranslate material-icons\\" aria-hidden=\\"true\\" role=\\"img\\">arrow_back</i></span></button></div>
            <div data-v-f3901ebe=\\"\\" class=\\"q-carousel__control absolute absolute-right carousel-nav-btn\\" style=\\"margin: 210px 16px;\\"><button data-v-f3901ebe=\\"\\" class=\\"q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--round bg-white text-black q-btn--actionable q-focusable q-hoverable q-btn--dense carousel-nav-btn\\" style=\\"font-size: 12px;\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><i class=\\"q-icon notranslate material-icons\\" aria-hidden=\\"true\\" role=\\"img\\">arrow_forward</i></span></button></div>
            <div data-v-f3901ebe=\\"\\" class=\\"q-carousel__control absolute absolute-bottom slide-counter-container\\" style=\\"margin: 18px;\\">
              <div data-v-f3901ebe=\\"\\" class=\\"slide-counter\\">1/3</div>
            </div>
          </div>
        </div>
        <div data-v-69d547a9=\\"\\" class=\\"drawer-section-title q-mb-md\\">Facilidades do hotel</div>
        <div data-v-69d547a9=\\"\\" class=\\"amenities-section q-mb-md\\">
          <div data-v-accd225c=\\"\\" data-v-69d547a9=\\"\\">
            <div data-v-accd225c=\\"\\" class=\\"row q-col-gutter-md\\">
              <div data-v-accd225c=\\"\\" class=\\"col-12 col-sm-6 col-md-3\\">
                <div data-v-accd225c=\\"\\" class=\\"row items-center q-gutter-x-xs text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">help_outline</i><span data-v-accd225c=\\"\\" class=\\"q-ml-xs text-caption\\">Wi-Fi</span></div>
              </div>
              <div data-v-accd225c=\\"\\" class=\\"col-12 col-sm-6 col-md-3\\">
                <div data-v-accd225c=\\"\\" class=\\"row items-center q-gutter-x-xs text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">restaurant_menu</i><span data-v-accd225c=\\"\\" class=\\"q-ml-xs text-caption\\">Café da manhã</span></div>
              </div>
              <div data-v-accd225c=\\"\\" class=\\"col-12 col-sm-6 col-md-3\\">
                <div data-v-accd225c=\\"\\" class=\\"row items-center q-gutter-x-xs text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">store</i><span data-v-accd225c=\\"\\" class=\\"q-ml-xs text-caption\\">Loja de presentes</span></div>
              </div>
              <div data-v-accd225c=\\"\\" class=\\"col-12 col-sm-6 col-md-3\\">
                <div data-v-accd225c=\\"\\" class=\\"row items-center q-gutter-x-xs text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">hot_tub</i><span data-v-accd225c=\\"\\" class=\\"q-ml-xs text-caption\\">Jacuzzi</span></div>
              </div>
            </div>
            <transition-stub data-v-accd225c=\\"\\" appear=\\"false\\" persisted=\\"false\\" css=\\"false\\">
              <div data-v-accd225c=\\"\\" class=\\"q-mt-md\\" style=\\"display: none;\\">
                <div data-v-accd225c=\\"\\" class=\\"row q-col-gutter-md\\">
                  <div data-v-accd225c=\\"\\" class=\\"col-12 col-sm-6 col-md-3\\">
                    <div data-v-accd225c=\\"\\" class=\\"row items-center q-gutter-x-xs text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">medical_services</i><span data-v-accd225c=\\"\\" class=\\"q-ml-xs text-caption\\">Serviço médico</span></div>
                  </div>
                  <div data-v-accd225c=\\"\\" class=\\"col-12 col-sm-6 col-md-3\\">
                    <div data-v-accd225c=\\"\\" class=\\"row items-center q-gutter-x-xs text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">local_library</i><span data-v-accd225c=\\"\\" class=\\"q-ml-xs text-caption\\">Biblioteca</span></div>
                  </div>
                  <div data-v-accd225c=\\"\\" class=\\"col-12 col-sm-6 col-md-3\\">
                    <div data-v-accd225c=\\"\\" class=\\"row items-center q-gutter-x-xs text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">golf_course</i><span data-v-accd225c=\\"\\" class=\\"q-ml-xs text-caption\\">Campo de golfe</span></div>
                  </div>
                  <div data-v-accd225c=\\"\\" class=\\"col-12 col-sm-6 col-md-3\\">
                    <div data-v-accd225c=\\"\\" class=\\"row items-center q-gutter-x-xs text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">surfing</i><span data-v-accd225c=\\"\\" class=\\"q-ml-xs text-caption\\">Esportes aquáticos</span></div>
                  </div>
                  <div data-v-accd225c=\\"\\" class=\\"col-12 col-sm-6 col-md-3\\">
                    <div data-v-accd225c=\\"\\" class=\\"row items-center q-gutter-x-xs text-grey-6\\"><i data-v-accd225c=\\"\\" class=\\"q-icon notranslate material-icons\\" style=\\"font-size: 18px;\\" aria-hidden=\\"true\\" role=\\"presentation\\">child_care</i><span data-v-accd225c=\\"\\" class=\\"q-ml-xs text-caption\\">Clube de crianças</span></div>
                  </div>
                </div>
              </div>
            </transition-stub>
            <div data-v-accd225c=\\"\\" class=\\"row justify-center q-mt-md\\"><button data-v-accd225c=\\"\\" class=\\"q-btn q-btn-item non-selectable no-outline q-btn--outline q-btn--rectangle q-btn--rounded text-light-blue-5 q-btn--actionable q-focusable q-hoverable\\" style=\\"font-size: 10px;\\" tabindex=\\"0\\" type=\\"button\\"><span class=\\"q-focus-helper\\"></span><span class=\\"q-btn__content text-center col items-center q-anchor--skip justify-center row\\"><span class=\\"block\\">Mostrar todas as facilidades</span></span></button></div>
          </div>
        </div>
        <div data-v-69d547a9=\\"\\" class=\\"drawer-section-title q-mb-md\\">Conheça um pouco mais</div>
        <div data-v-69d547a9=\\"\\" class=\\"text-grey-6\\">Hotel de Teste Description</div>
      </div>
    </div>
    <!---->
  </div>
  <!---->
  <div class=\\"q-scrollarea__bar q-scrollarea__bar--v absolute-right q-scrollarea__bar--invisible\\" aria-hidden=\\"true\\"></div>
  <div class=\\"q-scrollarea__bar q-scrollarea__bar--h absolute-bottom q-scrollarea__bar--invisible\\" aria-hidden=\\"true\\"></div>
  <div class=\\"q-scrollarea__thumb q-scrollarea__thumb--v absolute-right q-scrollarea__thumb--invisible\\" style=\\"top: 0px; height: 0px; right: 0px;\\" aria-hidden=\\"true\\"></div>
  <div class=\\"q-scrollarea__thumb q-scrollarea__thumb--h absolute-bottom q-scrollarea__thumb--invisible\\" style=\\"left: 0px; width: 0px; bottom: 0px;\\" aria-hidden=\\"true\\"></div>
</div>"
`;
